### DISCUSSION TITLE
LastChangeBody

### DISCUSSION THREAD

**Discussion by Mephisto090** [mcuiobroker/docuJarvis#11] [Q&A] [answered]
Hi mcu,

wo war im Doc noch mal zu finden das ich so ein Wert (2023-01-08T18:13:03.785Z) in "vor 14 Minuten" ändern kann?

VG :)

**Comment by mcuiobroker** [comment, marked as answer]
https://mcuiobroker.gitbook.io/jarvis-infos/jarvis-v3/besonderheiten-v3/geraet/datenpunkt-eigenschaft/utc-gmt-datum#differenz-zu-jetzt-anzeigen-distancetonow

**Reply by Mephisto090** [reply]
Das er aber wie bei Jarvis es automatisch macht geht hier glaube nicht oder?

Sprich:
"Vor 20 Sekunden"
"vor 12 Minuten"
"vor 1 Stunde"

**Reply by mcuiobroker** [reply]
Nein. Ginge irgendwie auch, aber zu aufwendig.

**Reply by Mephisto090** [reply]
Theoretisch eine kleine Abfrage mit if. Ich versuche mich mal. :)

**Reply by mcuiobroker** [reply]
Gerne, wenn du es geschafft hast, stell es rein, dann kommt es in die Doku.

**Reply by Mephisto090** [reply]
Theoretisch müsste das funktionieren. Mein Problem ist, ich kann es nicht testen da ich nicht weiß wie if/elseif da aussehen muss. Irgendwie akzeptiert er auch nicht den String :(

```
{
"value":"val => 
	{
		if (!val) 
		{
			return '';
		} 

		let ms = new Date().getTime()-(new Date(val.substr(0,4)+'-'+val.substr(5,2)+'-'+val.substr(8,2)+' '+val.substr(11,8)).getTime()); 
		let d, h, m, s;  
		s = Math.floor(ms / 1000);  
		m = Math.floor(s / 60);  
		s = s % 60;  
		h = Math.floor(m / 60);  
		m = m % 60;  
		d = Math.floor(h / 24);  
		h = h % 24; 

		if (d > 0) {
			if (d > 1) {
				return 'vor '+d+' Tagen';
			} else {
				return 'vor '+d+' Tag';
			}
		} else if (h > 0) {
			if (h > 1) {
				return 'vor '+h+' Stunden';
			} else {
				return 'vor '+h+' Stunde';
			}
		} else if (m > 0) {
			if (m > 1) {
				return 'vor '+m+' Minuten';
			} else {
				return 'vor '+m+' Minute';
			}
		} else {
			if (s > 1) {
				return 'vor '+s+' Sekunden';
			} else {
				return 'vor '+s+' Sekunde';
			}
		} 
	}"
}
```

**Reply by Mephisto090** [reply]
Es funktioniert:

![image](https://user-images.githubusercontent.com/85995176/211211917-e3221cac-f71e-4e09-b1fb-2603d0e85967.png)
![image](https://user-images.githubusercontent.com/85995176/211211947-fae74e4c-c9bb-4f80-b179-003a45965caf.png)


`{"value":"val => {if (!val) {return '';} let ms = new Date().getTime()-(new Date(val.substr(0,4)+'-'+val.substr(5,2)+'-'+val.substr(8,2)+' '+val.substr(11,8)).getTime()); let d, h, m, s;  s = Math.floor(ms / 1000);  m = Math.floor(s / 60);  s = s % 60;  h = Math.floor(m / 60);  m = m % 60;  d = Math.floor(h / 24);  h = h % 24; if (d > 0) {if (d > 1) { return 'vor '+d+' Tagen'; } else { return 'vor '+d+' Tag'; } } else if (h > 0) { if (h > 1) { return 'vor '+h+' Stunden'; } else { return 'vor '+h+' Stunde'; } } else if (m > 0) { if (m > 1) { return 'vor '+m+' Minuten'; } else { return 'vor '+m+' Minute'; } } else { if (s > 1) { return 'vor '+s+' Sekunden'; } else { return 'vor '+s+' Sekunde'; } } }"}`

man könnte jetzt natürlich noch kleinteiliger werden wie bei Jarvis. Was denkst du. Soll es so ähnlich sein wie bei Jarvis? Dies dürfte eigentlich nur alles im Sekundenbereich sein, was großartig geändert werden müsste.

aktualisiert sich leider nicht von alleine. Gibt es da eine Möglichkeit?  :(


Bekommt man das irgendwie auch da hin wo der LastChangeBody wäre?

**Reply by Mephisto090** [reply]
So wäre es ähnlich Jarvis:

```
{
"value":"val => 
	{
		if (!val) 
		{
			return '';
		} 

		let ms = new Date().getTime()-(new Date(val.substr(0,4)+'-'+val.substr(5,2)+'-'+val.substr(8,2)+' '+val.substr(11,8)).getTime()); 
		let d, h, m, s;  
		s = Math.floor(ms / 1000);  
		m = Math.floor(s / 60);  
		s = s % 60;  
		h = Math.floor(m / 60);  
		m = m % 60;  
		d = Math.floor(h / 24);  
		h = h % 24; 

		if (d > 0) { 
			if (d > 1) { 
				return 'vor '+d+' Tagen'; 
			} else { 
				return 'vor '+d+' Tag'; 
			} 
		} else if (h > 0) { 
			if (h > 1) { 
				return 'vor '+h+' Stunden'; 
			} else { 
				return 'vor '+h+' Stunde'; 
			} 
		} else if (m > 0) { 
			if (m > 45) { 
				return 'vor etwa 1 Stunde'; 
			} else if (m > 1) { 
				return 'vor '+m+' Minuten'; 
			} else { 
				return 'vor '+m+' Minute'; 
			} 
		} else { 
			if (s > 29) { 
				return 'vor weniger als 1 Minute';  
			} else if  (s > 19) { 
				return 'vor halben Minute';  
			} else if  (s > 10) { 
				return 'vor weniger als 20 Sekunden';  
			} else { 
				return 'vor weniger als 10 Sekunden'; 
			} 
		}  
	}" 
} 
```

`{"value":"val => {if (!val) {return '';} let ms = new Date().getTime()-(new Date(val.substr(0,4)+'-'+val.substr(5,2)+'-'+val.substr(8,2)+' '+val.substr(11,8)).getTime()); let d, h, m, s;  s = Math.floor(ms / 1000);  m = Math.floor(s / 60);  s = s % 60;  h = Math.floor(m / 60);  m = m % 60;  d = Math.floor(h / 24);  h = h % 24; if (d > 0) { if (d > 1) { return 'vor '+d+' Tagen'; } else { return 'vor '+d+' Tag'; } } else if (h > 0) { if (h > 1) { return 'vor '+h+' Stunden'; } else { return 'vor '+h+' Stunde'; } } else if (m > 0) { if (m > 45) { return 'vor etwa 1 Stunde'; } else if (m > 1) { return 'vor '+m+' Minuten'; } else { return 'vor '+m+' Minute'; } } else { if (s > 29) { return 'vor weniger als 1 Minute';  } else if  (s > 19) { return 'vor halben Minute';  } else if  (s > 10) { return 'vor weniger als 20 Sekunden';  } else { return 'vor weniger als 10 Sekunden'; } }  }" } 
`



Kannst ja gerne beides anbieten im Doc

**Reply by mcuiobroker** [reply]
Sehr schön beide aufgenommen.

**Reply by Mephisto090** [reply]
Wo hast es drin? Würde es mir gerne mal anschauen :)

**Comment by mcuiobroker** [comment]
Die Aktualisierung kannst du mal testen
```
let intRefresh = setInterval(async function(){
	refreshDeviceState("DeviceID-bbea1d7b-d2ba-4c1f-9afd-ee0ab7f60065", "DEINStateKey-Name"); 
},60000);
```
Unter scripts einfügen

**Reply by Mephisto090** [reply]
muss ich bei DeviceID noch was ändern? Was meinst du mit StateKey-Name?

**Reply by mcuiobroker** [reply]
Der Link von oben. DeviceId musst du anpassen.  Statekey ist der Name vom DP im Gerät.

**Reply by Mephisto090** [reply]
> Der Link von oben.

bin ich blind, bei mir sieht das aus wie gestern :D

**Reply by Mephisto090** [reply]
![image](https://user-images.githubusercontent.com/85995176/211367455-707121f1-5319-4e30-9e0c-5bc1a4a393b8.png)

**Reply by mcuiobroker** [reply]
Liegt am Browser oder Doku. STRG+F5 auf einem anderen Unterpunkt der Doku.

**Reply by Mephisto090** [reply]
> DeviceId musst du anpassen. Statekey ist der Name vom DP im Gerät.

ich habe ja mehrere Deviceses, muss ich die Funktion jetzt mehrmals einfügen?

**Reply by Mephisto090** [reply]
Kannst gerne noch bei "noch kleinteiliger" reinschreiben ähnlich Jarvis :)

**Reply by Mephisto090** [reply]
```
let intRefresh = setInterval(async function(){
	refreshDeviceState("279e828b-a5bd-4ce7-ba13-3ef328bd27cd_3a0a0", "lastUpdateTemperature");  /* Arbeitszimmer */
        refreshDeviceState("279e828b-a5bd-4ce7-ba13-3ef328bd27cd_3a0a0", "lastUpdateHumidity");  /* Arbeitszimmer */

	refreshDeviceState("279e828b-a5bd-4ce7-ba13-3ef328bd27cd_3a0a0_e98ad_0a418_e7be2", "lastUpdateTemperature");  /* Gäste-Bad */
        refreshDeviceState("279e828b-a5bd-4ce7-ba13-3ef328bd27cd_3a0a0_e98ad_0a418_e7be2", "lastUpdateHumidity");  /* Gäste-Bad */

	refreshDeviceState("279e828b-a5bd-4ce7-ba13-3ef328bd27cd_3a0a0_e98ad_fa44e", "lastUpdateTemperature");  /* Keller */
        refreshDeviceState("279e828b-a5bd-4ce7-ba13-3ef328bd27cd_3a0a0_e98ad_fa44e", "lastUpdateHumidity");  /* Keller */

	refreshDeviceState("279e828b-a5bd-4ce7-ba13-3ef328bd27cd_3a0a0_e98ad_fa44e_31f58", "lastUpdateTemperature");  /* Küche */
        refreshDeviceState("279e828b-a5bd-4ce7-ba13-3ef328bd27cd_3a0a0_e98ad_fa44e_31f58", "lastUpdateHumidity");  /* Küche  */

	refreshDeviceState("279e828b-a5bd-4ce7-ba13-3ef328bd27cd_3a0a0_e98ad", "lastUpdateTemperature");  /* Schlafzimmer */
        refreshDeviceState("279e828b-a5bd-4ce7-ba13-3ef328bd27cd_3a0a0_e98ad", "lastUpdateHumidity");  /* Schlafzimmer  */

	refreshDeviceState("279e828b-a5bd-4ce7-ba13-3ef328bd27cd_3a0a0_e98ad_0a418", "lastUpdateTemperature");  /* Veranda */
        refreshDeviceState("279e828b-a5bd-4ce7-ba13-3ef328bd27cd_3a0a0_e98ad_0a418", "lastUpdateHumidity");  /* Veranda  */
        refreshDeviceState("279e828b-a5bd-4ce7-ba13-3ef328bd27cd_3a0a0_e98ad_0a418", "lastUpdatePressure");  /* Veranda  */

	refreshDeviceState("279e828b-a5bd-4ce7-ba13-3ef328bd27cd", "lastUpdateTemperature");  /* Wohnzimmer */
        refreshDeviceState("279e828b-a5bd-4ce7-ba13-3ef328bd27cd", "lastUpdateHumidity");  /* Wohnzimmer  */
},5000);
```

nimm das ruhig noch in die Doku mit auf :) 


Bekommen wir das irgendwie auf die Linke Seite wo sonst das LastChangeBody wäre?

**Reply by Mephisto090** [reply]
Hier noch ein Bildchen für kleinteiliger

![image](https://user-images.githubusercontent.com/85995176/211375643-abe55cef-1eb8-4fca-a614-b141f681d160.png)

**Reply by mcuiobroker** [reply]
Läuft das jetzt mit dem Update der Daten?

**Comment by mcuiobroker** [comment]
Soll das rechts neben dem LastChangeBody? Sieht doch so gut aus?

**Reply by Mephisto090** [reply]
Ne ich dachte eigentlich Links unter Schlafzimmer, ...

**Reply by mcuiobroker** [reply]
Und dann den LastChangebody raus lassen?

**Reply by Mephisto090** [reply]
Ja genau, so war meine Idee. LastChangeBody zeigt ja nur an, wenn sich was am Wert geändert hat. Mit dem Update vom Gerät ist die Zeit manchmal eine andere.

**Comment by mcuiobroker** [comment]
Probiere es mal bitte hiermit:
https://mcuiobroker.gitbook.io/jarvis-infos/jarvis-v3/besonderheiten-v3/scripts-3.1.x-pro/statelist
Musst mal prüfen, ob es sich mit dem `refreshDeviceState` in die Quere kommt?
Könnte mir vorstellen, dass er da wieder text-align auf center setzt?
Dann muss man die Funktion nochmal anpassen.

**Reply by Mephisto090** [reply]
Ich habe es jetzt ganz einfach gemacht. Im Body State ist die LastUpdateTime drin und im Action State die Temepratur. So zeigt er vorne an, wann das Update war und hinten den Wert. :)
